File: Backend/RedditAffiliateEngine/Services/SerpCheckService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using RedditAffiliateEngine.Models;
using RedditAffiliateEngine.Repositories;

namespace RedditAffiliateEngine.Services
{
    /// <summary>
    /// Service for checking Google SERP (Search Engine Results Page) positions
    /// </summary>
    public class SerpCheckService : ISerpCheckService
    {
        private readonly IConfiguration _configuration;
        private readonly ILogger<SerpCheckService> _logger;
        private readonly HttpClient _httpClient;
        private readonly IRedditThreadRepository _threadRepository;
        private readonly ISerpResultRepository _serpResultRepository;
        private readonly string _serpApiKey;

        /// <summary>
        /// Constructor for SerpCheckService
        /// </summary>
        public SerpCheckService(
            IConfiguration configuration,
            ILogger<SerpCheckService> logger,
            HttpClient httpClient,
            IRedditThreadRepository threadRepository,
            ISerpResultRepository serpResultRepository)
        {
            _configuration = configuration;
            _logger = logger;
            _httpClient = httpClient;
            _threadRepository = threadRepository;
            _serpResultRepository = serpResultRepository;
            _serpApiKey = configuration["SERPAPI_KEY"] ?? throw new ArgumentNullException("SERPAPI_KEY is not configured");
        }

        /// <summary>
        /// Check the SERP position of a Reddit thread for a specific query
        /// </summary>
        public async Task<SerpCheckResponse> CheckSerpPositionAsync(SerpCheckRequest request)
        {
            try
            {
                _logger.LogInformation($"Checking SERP position for thread ID {request.ThreadId} with query: {request.Query}");
                
                // Get the thread
                var thread = await _threadRepository.GetThreadByIdAsync(request.ThreadId);
                if (thread == null)
                {
                    return new SerpCheckResponse
                    {
                        Success = false,
                        ThreadId = request.ThreadId,
                        Error = "Thread not found"
                    };
                }

                // Generate query if not provided
                string query = !string.IsNullOrEmpty(request.Query) 
                    ? request.Query 
                    : GenerateSearchQuery(thread);

                // Extract the domain part for checking in results
                string redditDomain = "reddit.com";
                string threadUrl = thread.Permalink;
                
                // Make sure the URL starts with reddit.com
                if (!threadUrl.Contains(redditDomain))
                {
                    threadUrl = $"{redditDomain}{threadUrl}";
                }

                // Check position
                int? position = await FindUrlPositionInSearchResultsAsync(query, threadUrl);

                // Save result to database
                var serpResult = new CreateSerpResultDto
                {
                    ThreadId = request.ThreadId,
                    Query = query,
                    Position = position,
                    IsRanked = position.HasValue
                };

                await _serpResultRepository.CreateSerpResultAsync(serpResult);

                // Update thread with SERP rank
                var threadUpdate = new UpdateRedditThreadDto
                {
                    SerpRank = position
                };
                
                await _threadRepository.UpdateThreadAsync(thread.Id, threadUpdate);

                return new SerpCheckResponse
                {
                    Success = true,
                    ThreadId = request.ThreadId,
                    Query = query,
                    Position = position
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error checking SERP position for thread ID {request.ThreadId}");
                return new SerpCheckResponse
                {
                    Success = false,
                    ThreadId = request.ThreadId,
                    Query = request.Query ?? "",
                    Error = ex.Message
                };
            }
        }

        /// <summary>
        /// Generate a search query based on a Reddit thread's content
        /// </summary>
        public string GenerateSearchQuery(RedditThread thread)
        {
            // Create a search query from the thread title
            // Remove special characters and limit to 60 chars
            string query = thread.Title;
            
            // If title is short, add some relevant words from the body
            if (query.Length < 30 && !string.IsNullOrEmpty(thread.Body))
            {
                string[] bodyWords = thread.Body
                    .Split(new[] { ' ', '\n', '\r', '\t' }, StringSplitOptions.RemoveEmptyEntries)
                    .Where(w => w.Length > 4)  // Only use words with > 4 chars
                    .Take(5)  // Take up to 5 words
                    .ToArray();
                
                if (bodyWords.Length > 0)
                {
                    query += " " + string.Join(" ", bodyWords);
                }
            }
            
            // Ensure the query is not too long
            if (query.Length > 60)
            {
                query = query.Substring(0, 60);
            }

            // Add the subreddit as additional context
            query += $" {thread.Subreddit} reddit";
            
            return query;
        }

        /// <summary>
        /// Check if a URL appears in search results for a given query
        /// </summary>
        public async Task<int?> FindUrlPositionInSearchResultsAsync(string query, string url, int maxPages = 5)
        {
            int position = 0;
            
            for (int page = 0; page < maxPages; page++)
            {
                int start = page * 10;
                string serpApiUrl = $"https://serpapi.com/search.json?engine=google&q={Uri.EscapeDataString(query)}&api_key={_serpApiKey}&start={start}";
                
                try
                {
                    _logger.LogInformation($"Querying SerpApi for: {query} (page {page+1})");
                    var response = await _httpClient.GetStringAsync(serpApiUrl);
                    
                    // Parse the response
                    var searchResults = JsonSerializer.Deserialize<JsonElement>(response);
                    
                    // Check if organic results exist
                    if (searchResults.TryGetProperty("organic_results", out var organicResults))
                    {
                        foreach (var result in organicResults.EnumerateArray())
                        {
                            position++;
                            
                            if (result.TryGetProperty("link", out var link))
                            {
                                string resultUrl = link.GetString() ?? "";
                                
                                // Check if the URL matches our target
                                if (resultUrl.Contains(url, StringComparison.OrdinalIgnoreCase))
                                {
                                    _logger.LogInformation($"Found match at position {position}: {resultUrl}");
                                    return position;
                                }
                            }
                        }
                    }
                    else
                    {
                        _logger.LogWarning("No organic results found in the SerpApi response");
                        break;
                    }
                    
                    // Check if there are more results
                    bool hasMore = false;
                    if (searchResults.TryGetProperty("pagination", out var pagination))
                    {
                        hasMore = pagination.TryGetProperty("next", out _);
                    }
                    
                    if (!hasMore)
                    {
                        break;
                    }
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, $"Error in SerpApi request for query: {query}");
                    throw;
                }
            }
            
            // URL not found in the search results
            return null;
        }
    }
}

File: server/services/serpCheckService.ts
import { SerpApi, GoogleSearch } from 'google-search-results-nodejs';
import { storage } from '../storage';

/**
 * Types for SERP checking
 */
export interface SerpCheckRequest {
  threadId: number;
  query?: string;
}

export interface SerpCheckResponse {
  success: boolean;
  threadId: number;
  query: string;
  position?: number | null;
  error?: string;
}

/**
 * Service for checking Google SERP (Search Engine Results Page) positions
 */
class SerpCheckService {
  private serpApiKey: string;
  private search: any; // SerpApi search instance

  constructor() {
    this.serpApiKey = process.env.SERPAPI_KEY || '';
    if (!this.serpApiKey) {
      console.error('SERPAPI_KEY is not configured in environment variables');
    }
    this.search = new GoogleSearch(this.serpApiKey);
  }

  /**
   * Check the SERP position of a Reddit thread for a specific query
   */
  async checkSerpPosition(request: SerpCheckRequest): Promise<SerpCheckResponse> {
    try {
      console.log(`Checking SERP position for thread ID ${request.threadId} with query: ${request.query}`);
      
      // Get the thread
      const thread = await storage.getThreadById(request.threadId);
      if (!thread) {
        return {
          success: false,
          threadId: request.threadId,
          query: request.query || '',
          error: 'Thread not found'
        };
      }

      // Generate query if not provided
      const query = request.query || this.generateSearchQuery(thread);

      // Extract the domain part for checking in results
      const redditDomain = 'reddit.com';
      let threadUrl = thread.permalink;
      
      // Make sure the URL starts with reddit.com
      if (!threadUrl.includes(redditDomain)) {
        threadUrl = `${redditDomain}${threadUrl}`;
      }

      // Check position
      const position = await this.findUrlPositionInSearchResults(query, threadUrl);

      // Save result to database
      const serpResult = {
        threadId: request.threadId,
        query,
        position,
        checkedAt: new Date(),
        isRanked: position !== null
      };

      await storage.createSerpResult(serpResult);

      // Update thread with SERP rank
      await storage.updateThread(thread.id, {
        serpRank: position
      });

      return {
        success: true,
        threadId: request.threadId,
        query,
        position
      };
    } catch (error: any) {
      console.error(`Error checking SERP position for thread ID ${request.threadId}:`, error);
      return {
        success: false,
        threadId: request.threadId,
        query: request.query || '',
        error: error.message
      };
    }
  }

  /**
   * Generate a search query based on a Reddit thread's content
   */
  generateSearchQuery(thread: any): string {
    // Create a search query from the thread title
    // Remove special characters and limit to 60 chars
    let query = thread.title;
    
    // If title is short, add some relevant words from the body
    if (query.length < 30 && thread.body) {
      const bodyWords = thread.body
        .split(/[\s\n\r\t]+/)
        .filter((w: string) => w.length > 4)  // Only use words with > 4 chars
        .slice(0, 5);  // Take up to 5 words
      
      if (bodyWords.length > 0) {
        query += ' ' + bodyWords.join(' ');
      }
    }
    
    // Ensure the query is not too long
    if (query.length > 60) {
      query = query.substring(0, 60);
    }

    // Add the subreddit as additional context
    query += ` ${thread.subreddit} reddit`;
    
    return query;
  }

  /**
   * Check if a URL appears in search results for a given query
   */
  async findUrlPositionInSearchResults(query: string, url: string, maxPages = 5): Promise<number | null> {
    let position = 0;
    
    for (let page = 0; page < maxPages; page++) {
      const start = page * 10;
      const params = {
        engine: 'google',
        q: query,
        start,
        api_key: this.serpApiKey
      };
      
      try {
        console.log(`Querying SerpApi for: ${query} (page ${page+1})`);
        
        // Use promise to get results from SerpApi
        const results = await new Promise((resolve, reject) => {
          this.search.json(params, (data: any) => {
            if (data.error) {
              reject(new Error(data.error));
            } else {
              resolve(data);
            }
          });
        });
        
        // Check if organic results exist
        const organicResults = (results as any).organic_results;
        if (organicResults && Array.isArray(organicResults)) {
          for (const result of organicResults) {
            position++;
            
            if (result.link && result.link.toLowerCase().includes(url.toLowerCase())) {
              console.log(`Found match at position ${position}: ${result.link}`);
              return position;
            }
          }
        } else {
          console.warn('No organic results found in the SerpApi response');
          break;
        }
        
        // Check if there are more results
        const pagination = (results as any).pagination;
        const hasMore = pagination && pagination.next;
        
        if (!hasMore) {
          break;
        }
      } catch (error) {
        console.error(`Error in SerpApi request for query: ${query}`, error);
        throw error;
      }
    }
    
    // URL not found in the search results
    return null;
  }
}

// Create a singleton instance
export const serpCheckService = new SerpCheckService();